R/repl.R

```diff
@@ -0,0 +1,204 @@
+#' Run a Python REPL
+#' 
+#' This function provides a Python REPL in the R session, which can be used
+#' to interactively run Python code. All code executed within the REPL is
+#' run within the Python 'main' module, and will remain accessible after the
+#' REPL is detached.
+#' 
+#' @param module An (optional) Python module to be imported before
+#'   the REPL is launched.
+#'   
+#' @param quiet Boolean; print a startup banner when launching the REPL? If
+#'   `FALSE`, the banner will be suppressed.
+#' 
+#' @export
+py_repl <- function(
+  module = NULL,
+  quiet = getOption("reticulate.repl.quiet", default = FALSE))
+{
+  # load module if requested
+  if (is.character(module))
+    import(module)
+  
+  # import other required modules for the REPL
+  builtins <- import_builtins(convert = FALSE)
+  main <- import_main(convert = FALSE)
+  codeop <- import("codeop", convert = TRUE)
+  
+  # grab references to the locals, globals of the main module
+  locals <- py_run_string("locals()")
+  globals <- py_run_string("globals()")
+  
+  # check to see if the current environment supports history
+  has_history <- tryCatch(
+    { utils::savehistory(tempfile()); TRUE },
+    error = function(e) FALSE
+  )
+  
+  if (has_history) {
+    
+    # if we have history, save and then restore the current R history
+    utils::savehistory()
+    on.exit(utils::loadhistory(), add = TRUE)
+    
+    # file to be used for command history during session
+    histfile <- getOption("reticulate.repl.histfile")
+    if (is.null(histfile))
+      histfile <- file.path(tempdir(), ".reticulatehistory")
+    
+    # load history (create emptu file if none exists yet)
+    if (!file.exists(histfile))
+      file.create(histfile)
+    utils::loadhistory(histfile)
+    
+  }
+  
+  # buffer of pending console input (we don't send input to
+  # the console until we have a complete Python statement)
+  buffer <- character()
+  
+  # command compiler (used to check if we've received a complete piece
+  # of Python input)
+  compiler <- codeop$CommandCompiler()
+  
+  # record whether the used has requested a quit
+  quit_requested <- FALSE
+  
+  # inform others that the reticulate REPL is active
+  options(reticulate.repl.active = TRUE)
```

Comment:
This could also set a mutable global in the reticulate package and then we could expose a `py_repl_active()` function. Would that work?

Reply:
We can do that. Any reason to prefer that over the R option?